// @atlaskit/avatar/dist/cjs/helpers/getPresenceSVG.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _taggedTemplateLiteral2 = require('babel-runtime/helpers/taggedTemplateLiteral');

var _taggedTemplateLiteral3 = _interopRequireDefault(_taggedTemplateLiteral2);

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _templateObject = (0, _taggedTemplateLiteral3.default)(['\n  fill: ', ';\n'], ['\n  fill: ', ';\n']);

exports.default = getPresenceSvg;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _styledComponents = require('styled-components');

var _styledComponents2 = _interopRequireDefault(_styledComponents);

var _theme = require('@atlaskit/theme');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var Svg = function Svg(props) {
  return _react2.default.createElement('svg', (0, _extends3.default)({
    height: '100%',
    version: '1.1',
    viewBox: '0 0 8 8',
    width: '100%',
    xmlns: 'http://www.w3.org/2000/svg'
  }, props));
};

var BusyCircle = _styledComponents2.default.circle(_templateObject, (0, _theme.themed)({ light: _theme.colors.R300, dark: _theme.colors.R200 }));
var BusyPath = _styledComponents2.default.path(_templateObject, _theme.colors.background);
var FocusPath = _styledComponents2.default.path(_templateObject, _theme.colors.purple);
var OfflineOuter = _styledComponents2.default.path(_templateObject, (0, _theme.themed)({ light: _theme.colors.N200, dark: _theme.colors.DN100 }));
var OfflineInner = _styledComponents2.default.path(_templateObject, (0, _theme.themed)({ light: _theme.colors.N40, dark: _theme.colors.DN500 }));
var OnlineCircle = _styledComponents2.default.circle(_templateObject, (0, _theme.themed)({ light: _theme.colors.G300, dark: _theme.colors.G200 }));

function getPresenceSvg(presence) {
  switch (presence) {
    case 'busy':
      return _react2.default.createElement(
        Svg,
        null,
        _react2.default.createElement(BusyCircle, { cx: '4', cy: '4', r: '4' }),
        _react2.default.createElement(BusyPath, { d: 'M3.3,1.9l2.8,2.8c0.2,0.2,0.2,0.5,0,0.7L5.4,6.1c-0.2,0.2-0.5,0.2-0.7,0L1.9,3.3c-0.2-0.2-0.2-0.5,0-0.7l0.7-0.7C2.8,1.7,3.1,1.7,3.3,1.9z' })
      );
    case 'focus':
      return _react2.default.createElement(
        Svg,
        null,
        _react2.default.createElement(FocusPath, { d: 'M4,8 C1.790861,8 0,6.209139 0,4 C0,1.790861 1.790861,0 4,0 C6.209139,0 8,1.790861 8,4 C8,6.209139 6.209139,8 4,8 Z M4,6.66666667 C5.47275933,6.66666667 6.66666667,5.47275933 6.66666667,4 C6.66666667,2.52724067 5.47275933,1.33333333 4,1.33333333 C2.52724067,1.33333333 1.33333333,2.52724067 1.33333333,4 C1.33333333,5.47275933 2.52724067,6.66666667 4,6.66666667 Z M4,5.33333333 C3.26362033,5.33333333 2.66666667,4.73637967 2.66666667,4 C2.66666667,3.26362033 3.26362033,2.66666667 4,2.66666667 C4.73637967,2.66666667 5.33333333,3.26362033 5.33333333,4 C5.33333333,4.73637967 4.73637967,5.33333333 4,5.33333333 Z' })
      );
    case 'offline':
      return _react2.default.createElement(
        Svg,
        null,
        _react2.default.createElement(OfflineOuter, { d: 'M4,8 C6.209139,8 8,6.209139 8,4 C8,1.790861 6.209139,0 4,0 C1.790861,0 0,1.790861 0,4 C0,6.209139 1.790861,8 4,8 Z M4,6 C5.1045695,6 6,5.1045695 6,4 C6,2.8954305 5.1045695,2 4,2 C2.8954305,2 2,2.8954305 2,4 C2,5.1045695 2.8954305,6 4,6 Z' }),
        _react2.default.createElement(OfflineInner, { d: 'M4,6 C5.1045695,6 6,5.1045695 6,4 C6,2.8954305 5.1045695,2 4,2 C2.8954305,2 2,2.8954305 2,4 C2,5.1045695 2.8954305,6 4,6 Z' })
      );
    case 'online':
      return _react2.default.createElement(
        Svg,
        null,
        _react2.default.createElement(OnlineCircle, { cx: '4', cy: '4', r: '4' })
      );
    default:
      return null;
  }
}
