// @atlaskit/avatar/dist/cjs/styled/Avatar.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.StatusWrapper = exports.PresenceWrapper = exports.Inner = undefined;

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _taggedTemplateLiteral2 = require('babel-runtime/helpers/taggedTemplateLiteral');

var _taggedTemplateLiteral3 = _interopRequireDefault(_taggedTemplateLiteral2);

var _templateObject = (0, _taggedTemplateLiteral3.default)(['\n  ', ';\n'], ['\n  ', ';\n']);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _styledComponents = require('styled-components');

var _styledComponents2 = _interopRequireDefault(_styledComponents);

var _constants = require('./constants');

var _utils = require('./utils');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

// OUTER WRAPPER
// eslint-disable-next-line react/prop-types
exports.default = function (_ref) {
  var children = _ref.children,
      size = _ref.size,
      stackIndex = _ref.stackIndex;

  var sizeValue = (0, _utils.getAvatarDimensions)({ size: size }, {
    includeBorderWidth: true,
    sizeOnly: true
  }) + 'px';
  return _react2.default.createElement(
    'div',
    {
      style: {
        display: 'inline-block',
        position: 'relative',
        outline: 0,
        width: sizeValue,
        height: sizeValue,
        zIndex: stackIndex
      }
    },
    children
  );
};

// INNER WRAPPER


var Inner = exports.Inner = _styledComponents2.default.div(_templateObject, _utils.getInnerStyles);

// PRESENCE WRAPPER


// TODO: do not have all icon sizes!
var getPresenceLayout = function getPresenceLayout(_ref2) {
  var appearance = _ref2.appearance,
      size = _ref2.size;

  var presencePosition = appearance === 'square' ? -(_constants.BORDER_WIDTH[size] * 2) : _constants.ICON_OFFSET[size];
  var presenceSize = _constants.ICON_SIZES[size];

  return {
    bottom: presencePosition + 'px',
    height: presenceSize + 'px',
    right: presencePosition + 'px',
    width: presenceSize + 'px'
  };
};

var PresenceWrapper = exports.PresenceWrapper = function PresenceWrapper(args) {
  return _react2.default.createElement(
    'span',
    {
      style: (0, _extends3.default)({
        pointerEvents: 'none',
        position: 'absolute'
      }, getPresenceLayout(args))
    },
    args.children
  );
};

// STATUS WRAPPER
var getStatusLayout = function getStatusLayout(_ref3) {
  var appearance = _ref3.appearance,
      size = _ref3.size;

  var statusPosition = appearance === 'square' ? _constants.SQUARE_ICON_OFFSET[size] : _constants.ICON_OFFSET[size];
  var statusSize = _constants.ICON_SIZES[size];

  return {
    height: statusSize + 'px',
    right: statusPosition + 'px',
    top: statusPosition + 'px',
    width: statusSize + 'px'
  };
};

// eslint-disable-next-line react/prop-types
var StatusWrapper = exports.StatusWrapper = function StatusWrapper(args) {
  return _react2.default.createElement(
    'span',
    {
      style: (0, _extends3.default)({
        position: 'absolute'
      }, getStatusLayout(args))
    },
    args.children
  );
};
