// @atlaskit/avatar/dist/cjs/styled/constants.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SQUARE_ICON_OFFSET = exports.ICON_OFFSET = exports.ICON_SIZES = exports.EXCESS_INDICATOR_FONT_SIZE = exports.BORDER_WIDTH = exports.AVATAR_RADIUS = exports.AVATAR_SIZES = exports.TRANSITION_DURATION = undefined;

var _theme = require('@atlaskit/theme');

var TRANSITION_DURATION = exports.TRANSITION_DURATION = '200ms';


var gridSizeValue = (0, _theme.gridSize)();

var AVATAR_SIZES = exports.AVATAR_SIZES = {
  xsmall: gridSizeValue * 2,
  small: gridSizeValue * 3,
  medium: gridSizeValue * 4,
  large: gridSizeValue * 5,
  xlarge: gridSizeValue * 12,
  xxlarge: gridSizeValue * 16
};

// border radius only applies to "square" avatars
var AVATAR_RADIUS = exports.AVATAR_RADIUS = {
  xsmall: 2,
  small: 2,
  medium: 3,
  large: 3,
  xlarge: 6,
  xxlarge: 12
};

var BORDER_WIDTH = exports.BORDER_WIDTH = {
  xsmall: 2,
  small: 2,
  medium: 2,
  large: 2,
  xlarge: 2,
  xxlarge: 2
};

// NOTE: sizes xsmall & xxlarge DO NOT support
// - groups
// - presence
// - status

var EXCESS_INDICATOR_FONT_SIZE = exports.EXCESS_INDICATOR_FONT_SIZE = {
  small: 10,
  medium: 11,
  large: 12,
  xlarge: 16
};

var ICON_SIZES = exports.ICON_SIZES = {
  small: 12,
  medium: 14,
  large: 15,
  xlarge: 18
};

var ICON_OFFSET = exports.ICON_OFFSET = {
  small: 0,
  medium: 0,
  large: 1,
  xlarge: 7
};

var SQUARE_ICON_OFFSET = exports.SQUARE_ICON_OFFSET = {
  small: 0,
  medium: 0,
  large: 0,
  xlarge: 1
};
