// @atlaskit/button/dist/cjs/styled/themeDefinitions.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _theme = require('@atlaskit/theme');

/**
 * Convert a hex colour code to RGBA.
 * @param {String} hex Hex colour code.
 * @param {Number} alpha Optional alpha value (defaults to 1).
 *
 */
/* eslint-disable no-bitwise */
var hex2rgba = function hex2rgba(hex) {
  var alpha = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : 1;

  var color = void 0;

  if (/^#([A-Fa-f0-9]{3}){1,2}$/.test(hex)) {
    color = hex.substring(1).split('');

    if (color.length === 3) {
      color = [color[0], color[0], color[1], color[1], color[2], color[2]];
    }

    color = '0x' + color.join('');

    // $FlowFixMe - `>>` oxperand can validly take a string value
    var r = color >> 16 & 255;
    // $FlowFixMe - `>>` operand can validly take a string value
    var g = color >> 8 & 255;
    // $FlowFixMe - `>>` operand can validly take a string value
    var b = color & 255;

    return 'rgba(' + [r, g, b].join(',') + ', ' + alpha + ')';
  }

  throw new Error('Bad Hex');
};
/* eslint-enable no-bitwise */

exports.default = {
  // Fallbacks
  fallbacks: {
    background: (0, _theme.themed)({ light: _theme.colors.N20A, dark: _theme.colors.DN70 }),
    color: (0, _theme.themed)({ light: _theme.colors.N400, dark: _theme.colors.DN400 }),
    textDecoration: 'none'
  },

  // Themes
  theme: {
    // Default appearance
    default: {
      background: {
        default: (0, _theme.themed)({ light: _theme.colors.N20A, dark: _theme.colors.DN70 }),
        hover: (0, _theme.themed)({ light: _theme.colors.N30A, dark: _theme.colors.DN60 }),
        active: (0, _theme.themed)({ light: hex2rgba(_theme.colors.B75, 0.6), dark: _theme.colors.B75 }),
        disabled: (0, _theme.themed)({ light: _theme.colors.N20A, dark: _theme.colors.DN70 }),
        selected: (0, _theme.themed)({ light: _theme.colors.N700, dark: _theme.colors.DN0 }),
        focusSelected: (0, _theme.themed)({ light: _theme.colors.N700, dark: _theme.colors.DN0 })
      },
      boxShadowColor: {
        focus: (0, _theme.themed)({ light: hex2rgba(_theme.colors.B200, 0.6), dark: _theme.colors.B75 }),
        focusSelected: (0, _theme.themed)({
          light: hex2rgba(_theme.colors.B200, 0.6),
          dark: _theme.colors.B75
        })
      },
      color: {
        default: (0, _theme.themed)({ light: _theme.colors.N400, dark: _theme.colors.DN400 }),
        active: (0, _theme.themed)({ light: _theme.colors.B400, dark: _theme.colors.B400 }),
        disabled: (0, _theme.themed)({ light: _theme.colors.N70, dark: _theme.colors.DN30 }),
        selected: (0, _theme.themed)({ light: _theme.colors.N20, dark: _theme.colors.DN400 }),
        focusSelected: (0, _theme.themed)({ light: _theme.colors.N20, dark: _theme.colors.DN400 })
      }
    },

    // Primary appearance
    primary: {
      background: {
        default: (0, _theme.themed)({ light: _theme.colors.B400, dark: _theme.colors.B100 }),
        hover: (0, _theme.themed)({ light: _theme.colors.B300, dark: _theme.colors.B75 }),
        active: (0, _theme.themed)({ light: _theme.colors.B500, dark: _theme.colors.B200 }),
        disabled: (0, _theme.themed)({ light: _theme.colors.N20A, dark: _theme.colors.DN70 }),
        selected: (0, _theme.themed)({ light: _theme.colors.N700, dark: _theme.colors.DN0 }),
        focusSelected: (0, _theme.themed)({ light: _theme.colors.N700, dark: _theme.colors.DN0 })
      },
      boxShadowColor: {
        focus: (0, _theme.themed)({ light: hex2rgba(_theme.colors.B200, 0.6), dark: _theme.colors.B75 }),
        focusSelected: (0, _theme.themed)({
          light: hex2rgba(_theme.colors.B200, 0.6),
          dark: _theme.colors.B75
        })
      },
      color: {
        default: (0, _theme.themed)({ light: _theme.colors.N0, dark: _theme.colors.DN30 }),
        disabled: (0, _theme.themed)({ light: _theme.colors.N70, dark: _theme.colors.DN30 }),
        selected: (0, _theme.themed)({ light: _theme.colors.N20, dark: _theme.colors.DN400 }),
        focusSelected: (0, _theme.themed)({ light: _theme.colors.N20, dark: _theme.colors.DN400 })
      }
    },

    // Warning appearance
    warning: {
      background: {
        default: (0, _theme.themed)({ light: _theme.colors.Y300, dark: _theme.colors.Y300 }),
        hover: (0, _theme.themed)({ light: _theme.colors.Y200, dark: _theme.colors.Y200 }),
        active: (0, _theme.themed)({ light: _theme.colors.Y400, dark: _theme.colors.Y400 }),
        disabled: (0, _theme.themed)({ light: _theme.colors.N20A, dark: _theme.colors.DN70 }),
        selected: (0, _theme.themed)({ light: _theme.colors.Y400, dark: _theme.colors.Y400 }),
        focusSelected: (0, _theme.themed)({ light: _theme.colors.Y400, dark: _theme.colors.Y400 })
      },
      boxShadowColor: {
        focus: (0, _theme.themed)({ light: _theme.colors.Y500, dark: _theme.colors.Y500 }),
        focusSelected: (0, _theme.themed)({ light: _theme.colors.Y500, dark: _theme.colors.Y500 })
      },
      color: {
        default: (0, _theme.themed)({ light: _theme.colors.N800, dark: _theme.colors.N800 }),
        disabled: (0, _theme.themed)({ light: _theme.colors.N70, dark: _theme.colors.DN30 }),
        selected: (0, _theme.themed)({ light: _theme.colors.N800, dark: _theme.colors.N800 }),
        focusSelected: (0, _theme.themed)({ light: _theme.colors.N800, dark: _theme.colors.N800 })
      }
    },

    // Danger appearance
    danger: {
      background: {
        default: (0, _theme.themed)({ light: _theme.colors.R400, dark: _theme.colors.R400 }),
        hover: (0, _theme.themed)({ light: _theme.colors.R300, dark: _theme.colors.R300 }),
        active: (0, _theme.themed)({ light: _theme.colors.R500, dark: _theme.colors.R500 }),
        disabled: (0, _theme.themed)({ light: _theme.colors.N20A, dark: _theme.colors.DN70 }),
        selected: (0, _theme.themed)({ light: _theme.colors.R500, dark: _theme.colors.R500 }),
        focusSelected: (0, _theme.themed)({ light: _theme.colors.R500, dark: _theme.colors.R500 })
      },
      boxShadowColor: {
        focus: (0, _theme.themed)({ light: _theme.colors.R100, dark: _theme.colors.R100 }),
        focusSelected: (0, _theme.themed)({ light: _theme.colors.R100, dark: _theme.colors.R100 })
      },
      color: {
        default: (0, _theme.themed)({ light: _theme.colors.N0, dark: _theme.colors.N0 }),
        disabled: (0, _theme.themed)({ light: _theme.colors.N70, dark: _theme.colors.DN30 }),
        selected: (0, _theme.themed)({ light: _theme.colors.N0, dark: _theme.colors.N0 }),
        focusSelected: (0, _theme.themed)({ light: _theme.colors.N0, dark: _theme.colors.N0 })
      }
    },

    // Help appearance
    help: {
      background: {
        default: (0, _theme.themed)({ light: _theme.colors.P400, dark: _theme.colors.P400 }),
        hover: (0, _theme.themed)({ light: _theme.colors.P200, dark: _theme.colors.P200 }),
        active: (0, _theme.themed)({ light: _theme.colors.P500, dark: _theme.colors.P500 }),
        disabled: (0, _theme.themed)({ light: _theme.colors.N20A, dark: _theme.colors.DN70 }),
        selected: (0, _theme.themed)({ light: _theme.colors.R500, dark: _theme.colors.R500 }),
        focusSelected: (0, _theme.themed)({ light: _theme.colors.R500, dark: _theme.colors.R500 })
      },
      boxShadowColor: {
        focus: (0, _theme.themed)({ light: _theme.colors.P100, dark: _theme.colors.P100 }),
        focusSelected: (0, _theme.themed)({ light: _theme.colors.P100, dark: _theme.colors.P100 })
      },
      color: {
        default: (0, _theme.themed)({ light: _theme.colors.N0, dark: _theme.colors.N0 }),
        disabled: (0, _theme.themed)({ light: _theme.colors.N70, dark: _theme.colors.DN30 }),
        selected: (0, _theme.themed)({ light: _theme.colors.N0, dark: _theme.colors.N0 }),
        focusSelected: (0, _theme.themed)({ light: _theme.colors.N0, dark: _theme.colors.N0 })
      }
    },

    // Link appearance
    link: {
      background: {
        default: (0, _theme.themed)({ light: 'none', dark: 'none' }),
        selected: (0, _theme.themed)({ light: _theme.colors.N700, dark: _theme.colors.N20 }),
        focusSelected: (0, _theme.themed)({ light: _theme.colors.N700, dark: _theme.colors.N20 })
      },
      boxShadowColor: {
        focus: (0, _theme.themed)({ light: hex2rgba(_theme.colors.B200, 0.6), dark: _theme.colors.B75 }),
        focusSelected: (0, _theme.themed)({
          light: hex2rgba(_theme.colors.B200, 0.6),
          dark: _theme.colors.B75
        })
      },
      color: {
        default: (0, _theme.themed)({ light: _theme.colors.B400, dark: _theme.colors.B100 }),
        hover: (0, _theme.themed)({ light: _theme.colors.B300, dark: _theme.colors.B75 }),
        active: (0, _theme.themed)({ light: _theme.colors.B500, dark: _theme.colors.B200 }),
        disabled: (0, _theme.themed)({ light: _theme.colors.N70, dark: _theme.colors.DN100 }),
        selected: (0, _theme.themed)({ light: _theme.colors.N20, dark: _theme.colors.N700 }),
        focusSelected: (0, _theme.themed)({ light: _theme.colors.N20, dark: _theme.colors.N700 })
      },
      textDecoration: {
        hover: 'underline'
      }
    },

    // Subtle appearance
    subtle: {
      background: {
        default: (0, _theme.themed)({ light: 'none', dark: 'none' }),
        hover: (0, _theme.themed)({ light: _theme.colors.N30A, dark: _theme.colors.DN60 }),
        active: (0, _theme.themed)({ light: hex2rgba(_theme.colors.B75, 0.6), dark: _theme.colors.B75 }),
        disabled: (0, _theme.themed)({ light: 'none', dark: 'none' }),
        selected: (0, _theme.themed)({ light: _theme.colors.N700, dark: _theme.colors.DN0 }),
        focusSelected: (0, _theme.themed)({ light: _theme.colors.N700, dark: _theme.colors.DN0 })
      },
      boxShadowColor: {
        focus: (0, _theme.themed)({ light: hex2rgba(_theme.colors.B200, 0.6), dark: _theme.colors.B75 }),
        focusSelected: (0, _theme.themed)({
          light: hex2rgba(_theme.colors.B200, 0.6),
          dark: _theme.colors.B75
        })
      },
      color: {
        default: (0, _theme.themed)({ light: _theme.colors.N400, dark: _theme.colors.DN400 }),
        active: (0, _theme.themed)({ light: _theme.colors.B400, dark: _theme.colors.B400 }),
        disabled: (0, _theme.themed)({ light: _theme.colors.N70, dark: _theme.colors.DN100 }),
        selected: (0, _theme.themed)({ light: _theme.colors.N20, dark: _theme.colors.DN400 }),
        focusSelected: (0, _theme.themed)({ light: _theme.colors.N20, dark: _theme.colors.DN400 })
      }
    },

    // Subtle Link appearance
    'subtle-link': {
      background: {
        default: (0, _theme.themed)({ light: 'none', dark: 'none' }),
        selected: (0, _theme.themed)({ light: _theme.colors.N700, dark: _theme.colors.N20 }),
        focusSelected: (0, _theme.themed)({ light: _theme.colors.N700, dark: _theme.colors.N20 })
      },
      boxShadowColor: {
        focus: (0, _theme.themed)({ light: hex2rgba(_theme.colors.B200, 0.6), dark: _theme.colors.B75 }),
        focusSelected: (0, _theme.themed)({
          light: hex2rgba(_theme.colors.B200, 0.6),
          dark: _theme.colors.B75
        })
      },
      color: {
        default: (0, _theme.themed)({ light: _theme.colors.N200, dark: _theme.colors.DN400 }),
        hover: (0, _theme.themed)({ light: _theme.colors.N90, dark: _theme.colors.B50 }),
        active: (0, _theme.themed)({ light: _theme.colors.N400, dark: _theme.colors.DN300 }),
        disabled: (0, _theme.themed)({ light: _theme.colors.N70, dark: _theme.colors.DN100 }),
        selected: (0, _theme.themed)({ light: _theme.colors.N20, dark: _theme.colors.DN400 }),
        focusSelected: (0, _theme.themed)({ light: _theme.colors.N20, dark: _theme.colors.DN400 })
      },
      textDecoration: {
        hover: 'underline'
      }
    }
  }
};
